### PR TITLE
[Runtime] Add fine-grained unified sequence parallel.

### PR DESCRIPTION
**Pull request by Thinkin999**
---
name: New Sequence Parallelism Support Request
about: Request support for a new sequence Parallelism in Galvatron
title: '[Parallelism] '
labels: 'parallelism-support'
assignees: ''

---

### Parallelism Description

- Parallelism Name: Zigzag Ring Flash Attention and Unified Sequence Parallel

- Paper/Reference:https://arxiv.org/abs/2310.01889

  ​           					https://arxiv.org/abs/2405.07719

- Original Implementation:https://github.com/zhuzilin/ring-flash-attention

  ​         									https://github.com/feifeibear/long-context-attention

#### Zigzag Ring Flash Attention

Zigzag Ring Flash Attention is a technique that seamlessly integrates the strengths of Ring Attention and Flash Attention to achieve highly efficient large-scale attention computations. 

- **Ring Attention Splitting and Global Computation: **The method begins by leveraging Ring Attention to partition the Query (Q), Key (K), and Value (V) matrices along the sequence dimension. These partitions are then processed through a ring-shaped communication pattern, enabling each GPU to collaboratively compute the global attention scores. 

- **Zigzag Data Transformation for Load Balancing: **To address the potential issue of uneven computational loads across GPUs, Zigzag Ring Flash Attention introduces a sophisticated Zigzag data transformation. This technique optimizes the way data is divided among GPUs, ensuring that each GPU receives a balanced share of the workload. 

- **Flash Attention for Memory Efficiency and Speed: **Building on the foundation of Ring Attention, each GPU employs the block-wise computation strategy of Flash Attention. This approach significantly reduces memory usage by breaking down the Q, K, and V matrices into smaller, manageable blocks. By avoiding the computation of large intermediate matrices and optimizing the softmax calculation, Flash Attention not only conserves memory but also accelerates the computation process.

#### Unified Sequence Parallel

Galvatron has integrated DeepSpeed-Ulysses sequence parallelism, and we have combined it with Zigzag Ring Flash Attention to form a hybrid parallelism approach called Unified Sequence Parallel (USP). This hybrid approach not only inherits the advantages of both methods but also overcomes their respective limitations.

DeepSpeed-Ulysses leverages All2All communication to handle segments of the Q (Query), K (Key), V (Value), and O (Output) tensors, shifting the tensor parallelism from the sequence dimension to the attention head dimension. This ensures that the computation for each attention head remains intact. On the other hand, Ring Attention employs nested double loops to coordinate communication and computation in a block-wise manner, obtaining the required K and V tensor segments through point-to-point communication.

However, both methods have their limitations: Ulysses sequence parallelism  is constrained by the number of attention heads, while Ring Attention's computational efficiency is reduced due to the subdivision of matrix multiplications. To address these limitations, USP divides the SP processes into two orthogonal groups: SP-Ring and SP-Ulysses, forming a two-dimensional grid structure. The algorithmic flow of USP includes the following steps:

1. Using AlltoAll communication to exchange Query, Key, and Value tensors.
2. Performing independent Ring-Attention computations within each group.
3. AlltoAll communication to ensure that each group has the complete Head Output results.

The benefits of USP are significant. It retains the capabilities of both original methods and adds additional advantages without any loss of functionality. USP overcomes the limitation of Ulysses' SP parallelism, and its hybrid communication pattern is more suitable for heterogeneous networks. In environments with PCIe and multi-node multi-GPU setups, USP demonstrates noticeable acceleration compared to using Ulysses or Ring Attention alone.

### Current Status

- [x] Zigzag Ring Flash Attention has been integrated into Galvatron.
- [x] Zigzag Ring Flash Attention can be used in a global mode (where each layer adopts the same parallel strategy) or in a fine-grained manner (where each layer can adopt different parallel strategies).
- [x] Zigzag Ring Flash Attention can be combined with DeepSpeed Ulysses to form Unified Sequence Parallel (USP).
- [x] Unified Sequence Parallel (USP) can be used in a global mode (where each layer adopts the same parallel strategy) or in a fine-grained manner (where each layer can adopt different parallel strategies).
- [x] RoPE (Rotary Position Embedding) has been adapted to work with both Zigzag Ring Flash Attention and Unified Sequence Parallel (USP) .

### Implementation Requirements

List specific requirements or challenges for implementing this parallelism in Galvatron:

​	1. Pre-integration of ZigZag Ring-Attention as the Communication Backbone

​	2. Fusion of ZigZag Ring-Attention with Ulysses SP

​	3. Inter-Layer Data Collection Strategy

 	4. Compatibility with Rotary Position Embedding (RoPE)

### Usage

First and foremost, it is important to note that Zigzag Ring Flash Attentionis applicable only to causal attention and must be used in conjunction with Flash Attention (by adding`--use-flash-attn`in the script,as shown in the examples below). We will first discuss the usage of Zigzag Ring Flash Attention and USP in Global Mode  (where the same parallel strategy is adopted for each layer), and then delve into the usage in Fine-grained Mode (where different parallel strategies can be adopted for each layer) under Galvatron JSON Mode.

Zigzag Ring Flash Attentioncan be combined with the following parallelisms to form hybrid parallelism:

• Megatron TP(Tensor Parallelism)

• Megatron SP(Sequence Parallelism,with`--sequence parallel`enabled)

• Data Parallelism

• Sharded Data Parallelism

• Pipeline Parallelism

The condition to be satisfied is: cp_size * tp_size * pp_size * dp_size = gpu_num.

Alternatively, it can be combined with Ulysses SP to form Unified Sequence Parallelism, which can further be combined with Data Parallelism, Sharded Data Parallelism,and Pipeline Parallelism. The condition remains the same: cp_size * tp_size * pp_size * dp_size = gpu_num.

Global Mode

##### Example 1:Using with Megatron TP,SP,and Other Parallelisms

Taking`Hetu-Galvatron/galvatron/models/llama_hf/scripts/train_dist.sh`as an example,we need to modify the`Parallel args`:

```shell
# Example
PARALLEL_ARGS="
    --pp_deg 1 \
    --global_tp_deg 1 \
    --global_tp_consec 1 \
    --sdp 0 \
    --global_checkpoint 1 \
    --vocab_tp 1 \
    --chunks 1 \
    --global_cp_deg 4 \
    --vocab_cp 4 \
    --pipeline_type pipedream_flush \
    --default_dp_type ddp \
    --mixed_precision bf16 \
    --use-flash-attn \
    --sequence-parallel  \
    --initialize_on_meta 1 "
```

Here,`--global_cp_deg `indicates the Context Parallel(CP)degree for the Transformer layer(Zigzag Ring Flash Attention is implemented as Context Parallel in Megatron,hence the abbreviation CP is used), and`--vocab_cp`indicates the sequence partitioning degree for the vocabulary embedding layer. In Global Mode,these two values are generally equal,but they can also be different. Note that`--use-flash-attn`must be included.

##### Example 2: Using with Ulysses SP and Other Parallelisms

When using Ulysses SP to replace Megatron TP and Megatron SP,and combining it with other parallelisms:

```shell
# Example
PARALLEL_ARGS="
    --pp_deg 1 \
    --global_tp_deg 2 \
    --global_tp_consec 1 \
    --sdp 0 \
    --global_checkpoint 1 \
    --vocab_tp 1 \
    --chunks 1 \
    --global_cp_deg 4 \
    --pipeline_type pipedream_flush \
    --default_dp_type ddp \
    --mixed_precision bf16 \
    --use-flash-attn \
    --sequence-parallel \
    --use-ulysses \
    --initialize_on_meta 1 "
```

When using Unified SP(USP) in Global Mode with Ulysses SP,add`--use-ulysses`and`--sequence-parallel`.The`global_tp_deg`will automatically be converted to`global_sp_deg`.

Fine-grained Mode (Galvatron JSON Mode)

When using Zigzag Ring Attention in Fine-grained Mode,we must add the`cp_sizes_enc`item in the JSON file and select an appropriate CP size for each layer. Similarly,choose an appropriate sequence partitioning dimension after `vcp`for the vocabulary:

```json
{
    "pp_deg": 1,
    "tp_sizes_enc": "2,4,8,2",
    "cp_sizes_enc": "2,2,1,4",
    "tp_consecutive_flags": "1,1,1,1",
    "use_sp": "0,1,1,0",
    "dp_types_enc": "1,1,1,1",
    "global_bsz": 8,
    "chunks": 1,
    "pp_division": "4",
    "checkpoint": "1,1,1,1",
    "pipeline_type": "pipedream_flush",
    "default_dp_type": "zero2",
    "vtp": 2,
    "vsp": 1,
    "vcp": 2
}
```

Note: Currently, we are working on modeling the computation time and memory usage of Zigzag Ring Flash Attention and USP. In the future,we will integrate Zigzag Ring Flash Attention and USP into the search dimensions of the Search Engine, enabling automatic search for the optimal fine-grained hybrid parallelism strategy under given conditions.

### Additional Context

If you have any questions, please contact with [liu.qingshuo@stu.pku.edu.cn](mailto:liu.qingshuo@stu.pku.edu.cn)

### CODE REVIEW

File: galvatron/core/runtime/tensor_parallel/transformer.py [hunk 1]
```diff
@@ -864,11 +883,14 @@ def forward(self, hidden_states, attention_mask, encoder_output=None, inference_
                 q, k, v = [
                     rearrange(x, "s b ... -> b s ...").contiguous() for x in (query_layer, key_layer, value_layer)
                 ]
-                if not self.sequence_parallel:
-                    with tensor_parallel.get_cuda_rng_tracker().fork():
-                        context_layer = self.core_attention_flash(q, k, v)
+                if self.use_zigzag_cp:
+                        context_layer = self.zigzag_ring_flash_attn(q, k, v)
                 else:
```
[the remark this answers is not in the document]
Comment (Thinkin999) [reply galvatron/core/runtime/tensor_parallel/transformer.py:888]: Copilot may not be right, but i nee to find  a way to make it easy to follow.
